reference/test/preconditioner/batch_isai_kernels.cpp

```diff
@@ -0,0 +1,409 @@
+                                         .on(exec);
+
+            for (size_t i = 0; i < umtxs.size(); i++) {
+                auto unbatch_prec = unbatch_prec_fact->generate(umtxs[i]);
+                unbatch_prec->apply(ub[i].get(), ux[i].get());
+            }
+
+        } else if (type == std::string("upper")) {
+            mtx = upper_tri_mtx;
+            umtxs = gko::test::share(mtx->unbatch());
+            batch_isai_type =
+                gko::preconditioner::batch_isai_input_matrix_type::upper_tri;
+
+            auto unbatch_prec_fact = unbatch_upper_prec_type::build()
+                                         .with_skip_sorting(true)
+                                         .with_sparsity_power(spy_power)
+                                         .on(exec);
+
+            for (size_t i = 0; i < umtxs.size(); i++) {
+                auto unbatch_prec = unbatch_prec_fact->generate(umtxs[i]);
+                unbatch_prec->apply(ub[i].get(), ux[i].get());
+            }
+
+
+        } else if (type == std::string("general")) {
+            mtx = general_mtx;
+            umtxs = gko::test::share(mtx->unbatch());
+            batch_isai_type =
+                gko::preconditioner::batch_isai_input_matrix_type::general;
+
+            auto unbatch_prec_fact = unbatch_gen_prec_type::build()
+                                         .with_skip_sorting(true)
+                                         .with_sparsity_power(spy_power)
+                                         .on(exec);
+
+            for (size_t i = 0; i < umtxs.size(); i++) {
+                auto unbatch_prec = unbatch_prec_fact->generate(umtxs[i]);
+                unbatch_prec->apply(ub[i].get(), ux[i].get());
+            }
+
+        } else {
+            GKO_NOT_IMPLEMENTED;
+        }
+
+
+        auto prec_fact = prec_type::build()
+                             .with_skip_sorting(true)
+                             .with_sparsity_power(spy_power)
+                             .with_isai_input_matrix_type(batch_isai_type)
+                             .on(exec);
+        auto prec = prec_fact->generate(mtx);
+        auto approx_inv = prec->get_const_approximate_inverse();
+
+        gko::kernels::reference::batch_isai::apply_isai(
+            exec, mtx.get(), approx_inv.get(), b.get(), x.get());
+
+        auto xs = x->unbatch();
+        for (size_t i = 0; i < umtxs.size(); i++) {
+            GKO_ASSERT_MTX_NEAR(ux[i], xs[i], r<value_type>::value);
+        }
+    }
+};
+
+TYPED_TEST_SUITE(BatchIsai, gko::test::ValueTypes);
+
+
+TYPED_TEST(BatchIsai, GeneralBatchIsaiGenerationIsEquivalentToUnbatchedWithSpy1)
+{
+    this->test_batch_isai_generation_is_eqvt_to_unbatched(
+        1, std::string("general"));
+}
+
+// TODO: Fix bug in normal isai
+// TYPED_TEST(BatchIsai,
+// GeneralBatchIsaiGenerationIsEquivalentToUnbatchedWithSpy2)
+// {
+//     this->test_batch_isai_generation_is_eqvt_to_unbatched(
+//         2, std::string("general"));
+// }
```

Comment:
Can you please log this in a Github issue, so that we make sure to fix it later ?

Reply:
Sure